### PR TITLE
Efficiently use in-order and out-of-order queues in the SYCL backends

### PR DESCRIPTION
**Pull request by uumesh**
* Adds the flexibility to use in-order out-of-order queues depending upon the host application.

* Default to in-order queues.

* Only order queue as necessary for efficiency (if it is out of order).

* Sets the same queue recursively through a hierarchy of Ceed objects (for the Ceed_parent and Ceed_delegate objects).

### CODE REVIEW

File: backends/sycl-gen/ceed-sycl-gen-operator-build.sycl.cpp
Comment (jeremylt) [inline backends/sycl-gen/ceed-sycl-gen-operator-build.sycl.cpp:752]:
```suggestion
  std::vector<sycl::event> e;
  
  if (!sycl_data->sycl_queue.is_in_order()) e = {sycl_data->sycl_queue.ext_oneapi_submit_barrier()};

  sycl::event copy_B       = sycl_data->sycl_queue.copy<Fields_Sycl>(&h_B, impl->B, 1, e);
  sycl::event copy_G       = sycl_data->sycl_queue.copy<Fields_Sycl>(&h_G, impl->G, 1, e);
  sycl::event copy_indices = sycl_data->sycl_queue.copy<FieldsInt_Sycl>(&h_indices, impl->indices, 1, e);
  
```
style issue I haven't figured out how to automate yet - we generally put a new line after variable declarations

Comment (uumesh) [reply backends/sycl-gen/ceed-sycl-gen-operator-build.sycl.cpp]: I have replaced it in this code segment. Do you want me to follow the same in every other similar code as well?

Comment (jeremylt) [reply backends/sycl-gen/ceed-sycl-gen-operator-build.sycl.cpp]: Yes please. One of these day's I'll figure out the way to automate this

Comment (uumesh) [reply backends/sycl-gen/ceed-sycl-gen-operator-build.sycl.cpp]: Done.

**Comment by jeremylt** [comment]
This all seems reasonable to me. I think all we need is the summary output of `make prove-all` on an env where these backends can run copied into this MR since we don't have automated CI for SYCL at this time.

File: backends/sycl-ref/ceed-sycl-ref-operator.sycl.cpp
Comment (jrwrigh) [inline backends/sycl-ref/ceed-sycl-ref-operator.sycl.cpp:836]:
```suggestion
    const CeedInt tid = idx % num_nodes;
    const CeedInt e   = idx / num_nodes;
```

**Comment by jrwrigh** [comment]
In addition to the above suggested change, I also had to apply this patch to get it to compile on Aurora with 2023.12.05.001 compilers:
```diff
diff --git i/backends/sycl-ref/ceed-sycl-ref.hpp w/backends/sycl-ref/ceed-sycl-ref.hpp
index 56544c38..fc7bc775 100644
--- i/backends/sycl-ref/ceed-sycl-ref.hpp
+++ w/backends/sycl-ref/ceed-sycl-ref.hpp
@@ -94,7 +94,7 @@ typedef struct {
 } CeedOperatorDiag_Sycl;

 typedef struct {
-  CeedInt     num_elem, block_size_x, block_size_y, elem_per_block;
+  CeedInt     num_elem, block_size_x, block_size_y, elems_per_block;
   CeedInt     num_e_mode_in, num_e_mode_out, num_qpts, num_nodes, block_size, num_comp;  // Kernel parameters
   bool        fallback;
   CeedScalar *d_B_in, *d_B_out;
```

I'm not sure why the above isn't showing up in the diff for the PR though, as the code on ALCF repo and the code here are definitely different.

**Comment by jrwrigh** [comment]
Testing on Aurora, `prove-all` shows everything passing except:
```
petsc-bps ................................ Failed 1/2 subtests
petsc-dmswarm ............................ ok
ex1-volume ............................... ok
ex2-surface .............................. ok
petsc-multigrid .......................... Failed 1/2 subtests
petsc-bpsswarm ........................... ok
solids-elasticity ........................ ok
fluids-navierstokes ...................... Failed 15/22 subtests

Test Summary Report
-------------------
fluids-py-smartsim_regression_framework (Wstat: 256 Tests: 0 Failed: 0)
  Non-zero exit status: 1
  Parse errors: No plan found in TAP output
petsc-bps                              (Wstat: 0 Tests: 2 Failed: 1)
  Failed test:  1
petsc-multigrid                        (Wstat: 0 Tests: 2 Failed: 1)
  Failed test:  2
fluids-navierstokes                    (Wstat: 0 Tests: 22 Failed: 15)
  Failed tests:  1, 3-4, 6-14, 17, 20-21
  ```
  
The PETSc failures are just due to the fact that I didn't have tetgen working, so it couldn't run the test with tetrahedral elements.

The fluids examples have some issue with segmentation faults, but I haven't had a chance to see where those faults are happening.

**Comment by jeremylt** [comment]
We should probably allow skipping tests that need tetgen in the test suite, like Ratel does

**Comment by jrwrigh** [comment]
Segmentation faults occur in the SYCL Kokkos implementation and are present on main as well. I'll go ahead and say that the PR is good to go with the code changes mentioned above.

**Review by jrwrigh — APPROVED** [review]

**Comment by jeremylt** [comment]
Yeah, that segfault seems related to PETSc sparse Kokkos matrices and not libCEED. Seems like we're good to merge once the two fixes above are applied.

**Comment by jeremylt** [comment]
Note, this PR will need to be updated (rebase or similar) due to minor updates in main

**Comment by jeremylt** [comment]
This looks mostly ready, minus some minor changes. Anything I can do to help get this merged?

**Comment by jeremylt** [comment]
Anything I can do to help this PR move? I worry about it going stale

**Comment by uumesh** [comment]
Is a rebase with main the only thing that needs to be done to get the PR merged?

**Comment by jeremylt** [comment]
A rebase and a clean test suite run is all that's needed

**Comment by jrwrigh** [comment]
@jeremylt Looks like gitlab CI isn't running on this branch for some reason. Obviously the functional testing needs to be done on Sunspot, but am a bit confused why it's not running here. Maybe since the branch is from a fork?

**Review by jrwrigh — APPROVED** [review]
All tests pass on Sunspot, except for some fluids tests. The failures are the same as reported [earlier](https://github.com/CEED/libCEED/pull/1456#issuecomment-1935051466) and are also occurring on `main`. So this is good as far as I'm concerned.

Given the failures, I'll wait until I hear back from @jedbrown and/or @jeremylt before merging.

**Comment by jeremylt** [comment]
Ready in my book

**Pull request merged**
